http/src/main/scala/org/http4s/blaze/http/http2/PingManager.scala

```diff
@@ -1,4 +1,58 @@
 package org.http4s.blaze.http.http2
 
-// Place holder
-trait PingManager
+import java.nio.ByteBuffer
+
+import org.log4s.getLogger
+
+import scala.concurrent.{Future, Promise}
+import scala.concurrent.duration.Duration
+
+private class PingManager(session: SessionCore) {
+
+  private[this] case class PingState(startedSystemTimeMs: Long, continuation: Promise[Duration])
+
+  private[this] val logger = getLogger
+  private[this] var currentPing: Option[PingState] = None
+
+  // Must be called from within the sessionExecutor
+  def ping(): Future[Duration] = currentPing match {
+    case Some(_) =>
+      val msg = "Ping already in progress"
+      val ex = new IllegalStateException(msg)
+      logger.warn(ex)(msg)
+      Future.failed(ex)
+
+    case None =>
+      val time = System.currentTimeMillis
+      val p = Promise[Duration]
+      currentPing = Some(PingState(time, p))
+      val data = new Array[Byte](8)
+      ByteBuffer.wrap(data).putLong(time)
+      val pingFrame = session.http2Encoder.pingFrame(data)
+      // TODO: we can do a lot of cool things with pings by managing when they are written
+      session.writeController.write(pingFrame)
+      logger.debug(s"Ping initiated at $time")
+      p.future
+  }
+
+  def pingAckReceived(data: Array[Byte]): Unit = {
+    currentPing match {
+      case None => // nop
+      case Some(PingState(sent, continuation)) =>
+        currentPing = None
+
+        if (ByteBuffer.wrap(data).getLong != sent) { // data guaranteed to be 8 bytes
+          val msg = "Received ping response with unknown data."
+          val ex = new Exception(msg)
+          logger.warn(ex)(msg)
+          continuation.tryFailure(ex)
+          ()
+        } else {
+          val duration = Duration.fromNanos((System.currentTimeMillis - sent) * 1000000)
```

Comment:
There's a `Duration.create(Long, TimeUnit)`.  I haven't checked the innards to see if that's faster or not.

Reply:
I bet its pretty quick: it appears to be an alias for the constructor. 😄 I'll definitely use `create`.